core/src/main/scala/shapeless/generic.scala

```diff
@@ -380,82 +381,85 @@ trait CaseClassMacros extends ReprTypes {
+          if (sym.isModuleClass) sym.toType
+          else appliedType(sym.toType, substituteArgs)
+        } else {
+          // Look for a path from the macro call site to the subtype.
+          val owner = sym.owner
+          val owners = ownerChain(if (isNamed) owner else owner.owner)
+          val prePaths = for (pre <- Iterator.iterate(basePre)(prefix).takeWhile(_ != NoPrefix))
+            yield (pre, owners.iterator.dropWhile(pre.baseType(_) == NoType))
+
+          // Find a path from a (sub-)prefix or the enclosing owner.
+          val (pre0, path) = prePaths.find(_._2.nonEmpty).getOrElse {
+            val enclosing = ownerChain(enclosingOwner)
+            val common = owners zip enclosing indexWhere { case (o1, o2) => o1 != o2 }
+            (NoPrefix, if (common < 0) Iterator.empty else owners.iterator drop common - 1)
           }
 
-          val suffix = ownerChain(sym).dropWhile(_ != basePre.typeSymbol)
-          val ctor =
-            if(suffix.isEmpty) {
-              if(sym.isModuleClass) {
-                val moduleSym = sym.asClass.module
-                val modulePre = prefix(moduleSym.typeSignatureIn(basePre))
-                c.internal.singleType(modulePre, moduleSym)
-              } else
-                appliedType(sym.toTypeIn(basePre), substituteArgs)
-            } else {
-              if(sym.isModuleClass) {
-                val path = suffix.tail.map(sym => normalizeTermName(sym.name))
-                val (modulePre, moduleSym) = mkDependentRef(basePre, path)
-                c.internal.singleType(modulePre, moduleSym)
-              } else if(isAnonOrRefinement(sym)) {
-                val path = suffix.tail.init.map(sym => normalizeTermName(sym.name))
-                val (valPre, valSym) = mkDependentRef(basePre, path)
-                c.internal.singleType(valPre, valSym)
-              } else {
-                val path = suffix.tail.init.map(sym => normalizeTermName(sym.name)) :+ suffix.last.name.toTypeName
-                val (subTpePre, subTpeSym) = mkDependentRef(basePre, path)
-                c.internal.typeRef(subTpePre, subTpeSym, substituteArgs)
-              }
-            }
-          if(!isAccessible(ctor))
-            abort(s"$tpe has an inaccessible subtype $ctor")
-          if(ctor <:< baseTpe) Some(ctor) else None
+          // Construct a stable prefix from the path.
+          val pre = path.drop(1).foldLeft(pre0) { (pre1, part) =>
+            if (part.isType) part.asType.toTypeIn(pre1)
+            else abort(s"$tpe has a subtype $sym with unstable prefix")
+          }
+
+          if (isNamed) {
+            if (sym.isModuleClass) sym.toTypeIn(pre)
+            else typeRef(pre, sym, substituteArgs)
+          } else {
+            def ownerIsSubType = owner.typeSignatureIn(pre) <:< baseTpe
+            if (owner.isTerm && owner.asTerm.isVal && ownerIsSubType) singleType(pre, owner)
+            else abort(s"$tpe has a subtype $sym with unstable prefix")
```

Comment:
The only change here is to check the owner of an anonymous class more strictly:
1. It should be a val (because `Generic` then uses its singleton type)
2. It should be a subtype of the base trait